### PR TITLE
TF: GPT-2 generation supports left-padding

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR does two things:
1. Enables left-padding with GPT-2 generation. 
    - It was working before only with XLA and was left as a TODO;
    - 🚨 Naturally, tests had to be changed. In the batched tests, the shortest sequence is now different, as a consequence of the correct processing of the left-padding;
    - Because we now have non-XLA left-padding, the XLA/non-XLA equivalence tests for GPT-2 now have two entries with different lengths;
    - An additional test was added, to ensure the output is the same regardless of left-padding.
2. Fix minor issues and TODOs in TF generate. In particular, I'd highlight the following:
    - All generated arrays are initialized with the `pad_token_id`, as opposed to with `0`. This was already present in `beam_search`, as one test caught it there;
    - Corrects the number of iterations in greedy search and sample -- it was one iteration short and resulting in outputs with `max_length-1` when length was the constraint (also an argument in favor by contradiction: if this change resulted in too many iterations, we would be attempting to write out of bounds of the `TensorArray`, which isn't the case)
    
___________________
Locally run slow tests: GPT-2, T5, BART, RAG, Encoder_Decoder, Vision_Encoder_Decoder, Speech2Text

### CODE REVIEW

File: src/transformers/generation_tf_utils.py [hunk 1]
```diff
@@ -1973,36 +1974,35 @@ def greedy_search(
...
             dynamic_size=False,
             size=max_length,
             clear_after_read=False,
         )
+        for i in range(max_length):
+            generated = generated.write(i, tf.broadcast_to(pad_token_id, (batch_size,)))
 
         # write prompt to generated
-        for i in range(seq_length):
+        for i in range(cur_len):
```
Comment (gante) [inline src/transformers/generation_tf_utils.py:1991]: `cur_len` is the most common variable name for the current length

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -2029,13 +2029,8 @@ def greedy_search_body_fn(generated, finished_sequences, next_tokens, current_po
                     decoder_hidden_states.append(outputs.hidden_states)
 
             # pre-process distribution
-            # TODO(pvp, joao, matt) - all the logits processors need to be adapted
-            # to be XLA compatible
-            input_ids = None
-            if not use_xla:
-                input_ids = tf.reshape(generated.concat(), (-1, batch_size))
-                input_ids = tf.transpose(input_ids[: current_pos[0]])
```
Comment (gante) [inline src/transformers/generation_tf_utils.py:2037]: no need to slice `input_ids`, as the updated `logits_processor` is expecting as input the fixed-length input

File: src/transformers/generation_tf_utils.py [hunk 3]
```diff
@@ -1973,36 +1974,36 @@ def greedy_search(
...
         generated = tf.TensorArray(
             element_shape=(batch_size,),
             dtype=tf.int32,
             dynamic_size=False,
             size=max_length,
             clear_after_read=False,
         )
+        if pad_token_id:  # ignores the cases when it is 0 or None
+            for i in range(max_length):
+                generated = generated.write(i, tf.broadcast_to(pad_token_id, (batch_size,)))
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_utils.py:1989]: Since `TensorArray` is write-once, will this not make it impossible to write anything more into `generated`?

Comment (gante) [reply src/transformers/generation_tf_utils.py:1989]: Interesting! It does say that in the docs, right at the top, but I've seen no limitations with this -- beam search already had this change 🤔 

I can also move this loop to populate any remaining positions with the pad token, it's probably better! WDYT?

File: src/transformers/generation_tf_utils.py [hunk 4]
```diff
@@ -2250,30 +2245,32 @@ def sample(
...
             element_shape=(batch_size,),
             dtype=tf.int32,
             dynamic_size=False,
             size=max_length,
             clear_after_read=False,
         )
-        finished_sequences = tf.zeros((batch_size,), dtype=tf.bool)
+        if pad_token_id:  # ignores the cases when it is 0 or None
+            for i in range(max_length):
+                generated = generated.write(i, tf.broadcast_to(pad_token_id, (batch_size,)))
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_utils.py:2260]: Same here, this looks like `generated` will just be a read-only array of `pad_token_id` after this!

File: src/transformers/generation_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py:1989]: Ah! Let me run some experiments, then. Maybe if you do this in a graph that never tries to backprop through the array, it lets you do it?

Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py:1989]: Tested and the docs are just entirely wrong, even for `tf.float32` arrays, at least when you're not in a `GradientTape` context!

File: src/transformers/generation_tf_utils.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py:2260]: Ignore this - resolved as per the conversation above.

Comment (gante) [reply src/transformers/generation_tf_utils.py:2260]: (see comment above :P)

**Review by Rocketknight1 — APPROVED** [review]
The code changes look straightforward, so I'm happy to merge this once the tests are green!

File: src/transformers/generation_tf_utils.py [hunk 5]
```diff
@@ -1653,12 +1655,25 @@ def _expand_to_num_beams(tensor: tf.Tensor, num_beams: int) -> tf.Tensor:
...
+        # however, they can also have left-padding (we definitely want to pick up on those)
+        has_trailing_pad_tokens = tf.math.reduce_any(inputs[:, 0] == pad_token_id)
+
         # Check if input is input_ids and padded -> only then is attention_mask defined
-        if is_input_ids and is_pad_token_in_inputs:
+        if (
+            is_input_ids
+            and is_pad_token_in_inputs
+            and (has_trailing_pad_tokens or is_pad_token_not_equal_to_eos_token_id)
+        ):
```
Comment (gante) [inline src/transformers/generation_tf_utils.py:1676]: @patrickvonplaten I think I found a bug in PT generation.

The `is_pad_token_not_equal_to_eos_token_id` condition exists in PT, it is copied here. It is used to infer the attention mask correctly when `eos_token_id==pad_token_id`, so that the eos doesn't get masked. However, GPT-2 also uses left-padding for generation (and we set `pad_token_id=eos_token_id`, because the former is missing). Without the new `has_trailing_pad_tokens`, the infered attention mask is incorrect when left-padding is present.

An example running from this branch, where TF passes the check and PT doesn't:
```
from transformers import GPT2Tokenizer, TFGPT2LMHeadModel, GPT2LMHeadModel

tokenizer = GPT2Tokenizer.from_pretrained("distilgpt2")
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "left"


tf_model = TFGPT2LMHeadModel.from_pretrained("distilgpt2")

tf_inputs = tokenizer(["Hello"], padding=True, return_tensors="tf")
tf_generation_output = tf_model.generate(tf_inputs.input_ids)
tf_outputs = tokenizer.batch_decode(tf_generation_output, skip_special_tokens=True)

tf_inputs_left_padded = tokenizer(["Hello", "Long input to trigger left padding"], padding=True, return_tensors="tf")
tf_generation_output_left_padded = tf_model.generate(tf_inputs_left_padded.input_ids)
tf_outputs_left_padded = tokenizer.batch_decode(tf_generation_output_left_padded, skip_special_tokens=True)
assert(tf_outputs[0][:60] == tf_outputs_left_padded[0][:60])


pt_model = GPT2LMHeadModel.from_pretrained("distilgpt2")

pt_inputs = tokenizer(["Hello"], padding=True, return_tensors="pt")
pt_generation_output = pt_model.generate(pt_inputs.input_ids)
pt_outputs = tokenizer.batch_decode(pt_generation_output, skip_special_tokens=True)

pt_inputs_left_padded = tokenizer(["Hello", "Long input to trigger left padding"], padding=True, return_tensors="pt")
pt_generation_output_left_padded = pt_model.generate(pt_inputs_left_padded.input_ids)
pt_outputs_left_padded = tokenizer.batch_decode(pt_generation_output_left_padded, skip_special_tokens=True)
assert(pt_outputs[0][:60] == pt_outputs_left_padded[0][:60])
```

Shall I also propagate the change to PT?

Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:1676]: Let's discuss in the call quickly :-)

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 6]
```diff
@@ -813,24 +813,21 @@ def get_output_embeddings(self):
     def set_output_embeddings(self, value):
         self.set_input_embeddings(value)
 
-    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, use_xla=False, **kwargs):
-        # TODO: (Joao) after the TF generator is complete, update GPT2 TF generation to match PT's. NB -- some GPT2
-        # tests will need to be fixed after the change
-
+    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:816]: great!

File: src/transformers/generation_tf_utils.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:1676]: As discussed let's rather throw a warning here :-)

**Comment by gante** [comment]
@patrickvonplaten made the changes we talked about 👍 There are a couple of tests failing, but I don't think they are related to these changes (like torch tests) 🤔

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Cool ! Looks good to me, if possible it'd be great if @sgugger could take a quick look here since the discussed logic of how to handle "automatic creation" of the `attention_mask` is a bit universal in Transformers.

**Comment by patrickvonplaten** [comment]
@sgugger, in short we have the following situation for the automatic attention mask creation. 

We **never** do this in the forward pass, but since a long time we have such a feature implementation implemented in PyTorch's generate: https://github.com/huggingface/transformers/blob/d156898f3b9b2c990e5963f5030a7143d57921a2/src/transformers/generation_utils.py#L490

So we won't be able to change that back in PyTorch (except for a major version). For now we do the following in PyTorch which handles the attention_mask creation correctly in 99% of the cases:
- If user doesn't provide the attention_mask **and** the padding token is in the input_ids **and** the padding token is not equal to the EOS token, we create an attention_mask automatically

This doesn't cover the edge-case where the user forwards both padding tokens and eos tokens + they are the same. Think the edge-case is really an edge case, but overall we should nudge the user to **always** provide an attention_mask if doing generate in batches. 

=> As a conclusion, we've now copied the PT logic 1-to-1 to TF generate & added a warning. After this PR is merged we should also add this warning to PT IMO. 

Does this sound good to you?

**Comment by patrickvonplaten** [comment]
Related: https://github.com/huggingface/transformers/pull/17444

**Comment by sgugger** [comment]
Sounds good to me @patrickvonplaten !

**Comment by patrickvonplaten** [comment]
Feel free to merge whenever @gante !

**Comment by ydshieh** [comment]
@gante King of TF `generate`!

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 7]
```diff
@@ -813,32 +813,29 @@ def get_output_embeddings(self):
...
-        if use_xla:
-            attention_mask = kwargs.get("attention_mask", None)
-            if past is not None and attention_mask is not None:
-                position_ids = tf.reduce_sum(attention_mask, axis=1, keepdims=True) - 1
-            elif attention_mask is not None:
-                position_ids = tf.math.cumsum(attention_mask, axis=1, exclusive=True)
+        if attention_mask is not None and position_ids is None:
+            position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
+            if past:
+                position_ids = tf.expand_dims(position_ids[:, -1], -1)
```
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:830]: @gante 

I think this block should be 
```
        if attention_mask is not None and position_ids is None:
            position_ids = tf.math.cumsum(attention_mask, axis=-1, exclusive=True)
        if past and position_ids is not None:
            position_ids = tf.expand_dims(position_ids[:, -1], -1)
```
This is not a real issue, as I don't think `prepare_inputs_for_generation` is ever called with `position_ids`.

However, to make the logic more consistent, it's better to have processing similiar to `inputs` and `token_type_ids` (i.e. explicitly cut the sequence)

Don't feel urgent for this though.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 8]
```diff
@@ -813,32 +813,29 @@ def get_output_embeddings(self):
     def set_output_embeddings(self, value):
         self.set_input_embeddings(value)
 
-    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, use_xla=False, **kwargs):
-        # TODO: (Joao) after the TF generator is complete, update GPT2 TF generation to match PT's. NB -- some GPT2
-        # tests will need to be fixed after the change
-
+    def prepare_inputs_for_generation(self, inputs, past=None, use_cache=None, **kwargs):
+        token_type_ids = kwargs.get("token_type_ids", None)
```
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:817]: @gante 

We will need the same thing as
https://github.com/gante/transformers/blob/ed266959eb9f73d13661320d8534f9a5628b2fd0/src/transformers/generation_utils.py#L616-L619

for `_update_model_kwargs_for_generation` in `generation_tf_utils`, after allowing `token_type_ids` in `prepare_inputs_for_generation` here.

Otherwise, it won't either get the correct length (if `past` is `None`).

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 9]
```diff
@@ -813,32 +813,29 @@ def get_output_embeddings(self):
...
         # only last token for inputs_ids if past is defined in kwargs
         if past:
             inputs = tf.expand_dims(inputs[:, -1], -1)
+            if token_type_ids is not None:
+                token_type_ids = tf.expand_dims(token_type_ids[:, -1], -1)
+
+        position_ids = kwargs.get("position_ids", None)
```
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:824]: I know this is also in PT-GPT2's `prepare_inputs_for_generation`, but I think it is not super perfect here. As we don't do anything for `position_ids` in `_update_model_kwargs_for_generation`, it would have some issue if `position_ids` is provided in `model_kwargs`.

However, it's very unlikely we have use cases where `position_ids` is provided explicitly for `generate`. So there is no urgency at all to address this here.

cc @patrickvonplaten 

p.s. In #17479, `https://github.com/huggingface/transformers/blob/9089b7b95a1b12f19b65872323f13f1f68a6eaa7/src/transformers/models/bart/modeling_tf_bart.py#L1400` is always created.

I believe we can always safely create `position_ids` here for (PT/TF)-GPT2 too ..?

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 7 above)
Comment (ydshieh) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:830]: Sorry, after checking `GPT2LMHeadModel`, I believe it should be 
```
        if attention_mask is not None and position_ids is None:
            # create position_ids on the fly for batch generation
            ...
        else:
            position_ids = None
```
i.e. we should set `position_ids = None` in an `else` case.
The reason is that `position_ids` is never get updated in `update_model_kwargs_for_generation`.
(Although I don't really think `position_ids` will ever be provided to `model_kwargs`  for `generate`)

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 10]
```diff
@@ -813,32 +813,29 @@ def get_output_embeddings(self):
...
         return {
             "input_ids": inputs,
             "attention_mask": attention_mask,
             "position_ids": position_ids,
             "past": past,
             "use_cache": use_cache,
+            "token_type_ids": token_type_ids,
         }
 
     def _update_model_kwargs_for_xla_generation(self, outputs, model_kwargs, current_pos, max_length):
```
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:841]: I think we also have to take care of `token_type_ids` in `_update_model_kwargs_for_xla_generation`. But it's not the most priority thing. Maybe putting a TODO comment here for now?
